akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala

```diff
@@ -0,0 +1,308 @@
+          deliverTo ! Delivery(pid, seqNr, msg, context.self)
+          waitingForConfirmation(
+            s.copy(producer = seqMsg.producer, receivedSeqNr = seqNr, requestedSeqNr = s.requestedSeqNr),
+            first)
+        } else if (seqNr > expectedSeqNr) {
+          logIfChangingProducer(s.producer, seqMsg, pid, seqNr)
+          context.log.infoN("from producer [{}], missing [{}], received [{}]", pid, expectedSeqNr, seqNr)
+          if (resendLost) {
+            seqMsg.producer ! Resend(fromSeqNr = expectedSeqNr)
+            resending(s.copy(producer = seqMsg.producer))
+          } else {
+            deliverTo ! Delivery(pid, seqNr, msg, context.self)
+            waitingForConfirmation(s.copy(producer = seqMsg.producer, receivedSeqNr = seqNr), first)
+          }
+        } else { // seqNr < expectedSeqNr
+          context.log.infoN("from producer [{}], deduplicate [{}], expected [{}]", pid, seqNr, expectedSeqNr)
+          Behaviors.same
+        }
+
+      case RetryRequest ⇒
+        retryRequest(s)
+
+      case Confirmed(seqNr) ⇒
+        context.log.warn("Unexpected confirmed [{}]", seqNr)
+        Behaviors.unhandled
+
+      case Start(_) =>
+        Behaviors.unhandled
+    }
+  }
+
+  // It has detected a missing seqNr and requested a Resend. Expecting a SequencedMessage from the
+  // ProducerController with the missing seqNr. Other SequencedMessage with different seqNr will be
+  // discarded since they were in flight before the Resend request and will anyway be sent again.
+  private def resending(s: State): Behavior[InternalCommand] = {
+    Behaviors.receiveMessage {
+      case seqMsg @ SequencedMessage(pid, seqNr, msg: A @unchecked, first) ⇒
+        checkProducerId(producerId, pid, seqNr)
+
+        // FIXME is SequencedMessage.first possible here?
+        if (seqNr == s.receivedSeqNr + 1) {
+          logIfChangingProducer(s.producer, seqMsg, pid, seqNr)
+          context.log.info("from producer [{}], received missing [{}]", pid, seqNr)
+          deliverTo ! Delivery(pid, seqNr, msg, context.self)
+          waitingForConfirmation(s.copy(producer = seqMsg.producer, receivedSeqNr = seqNr), first)
+        } else {
+          context.log.infoN("from producer [{}], ignoring [{}], waiting for [{}]", pid, seqNr, s.receivedSeqNr + 1)
+          Behaviors.same // ignore until we receive the expected
+        }
+
+      case RetryRequest ⇒
+        // in case the Resend message was lost
+        context.log.info("retry Resend [{}]", s.receivedSeqNr + 1)
+        s.producer ! Resend(fromSeqNr = s.receivedSeqNr + 1)
+        Behaviors.same
+
+      case Confirmed(seqNr) ⇒
+        context.log.warn("Unexpected confirmed [{}]", seqNr)
+        Behaviors.unhandled
+
+      case Start(_) =>
+        Behaviors.unhandled
+    }
+  }
+
+  // The message has been delivered to the consumer and it is now waiting for Confirmed from
+  // the consumer. New SequencedMessage from the ProducerController will be stashed.
```

Comment:
No, between PC and CC it's a demand window, not one-by-one.
The Request from CC to PC defines to the upper seqNr that the CC would like, PC can continue sending until reaching that limit. New Request is sent from CC after window/2.

Between P and PC it's one-by-one because that's easier for end user to reason about and that is quick local messages. Same, one-by-one between CC and C.